src/web/web_sockets.cpp

```diff
@@ -0,0 +1,560 @@
+int mg_ssl_if_mbed_random(void* connection, unsigned char* buffer,
+    size_t length)
+{
+    bc::data_chunk data;
+    data.reserve(length);
+    bc::pseudo_random_fill(data);
+
+    std::memcpy(buffer, data.data(), data.size());
+    return length;
+}
+
+}
+
+namespace libbitcoin {
+namespace server {
+
+using namespace asio;
+static constexpr auto poll_interval = 10u;
+
+struct mg_serve_http_opts web_socket_options;
+
+static int is_websocket(const mg_connection_ptr connection)
+{
+    return connection->flags & MG_F_IS_WEBSOCKET;
+}
+
+static std::string get_connection_address(mg_connection_ptr connection)
+{
+    static constexpr size_t max_address_length = 32;
+
+    std::array<char, max_address_length> address{};
+
+    mg_sock_addr_to_str(
+        &connection->sa, address.data(), address.size() - 1,
+        MG_SOCK_STRINGIFY_IP | MG_SOCK_STRINGIFY_PORT);
+
+    return std::string(address.data());
+}
+
+static void web_event_handler(mg_connection_ptr connection,
+    int event, void* data)
+{
+    switch(event)
+    {
+        case MG_EV_WEBSOCKET_HANDSHAKE_DONE:
+        {
+            auto web_sockets =
+                static_cast<bc::server::web_sockets*>(
+                    connection->user_data);
+
+            web_sockets->register_connection(connection);
+
+            LOG_DEBUG(LOG_SERVER)
+                << "Websocket client connection established ["
+                << get_connection_address(connection) << "] ("
+                << web_sockets->connection_count() << ")";
+
+            break;
+        }
+
+        case MG_EV_WEBSOCKET_FRAME:
+        {
+            const auto message =
+                static_cast<struct websocket_message*>(data);
+
+            const auto input = std::string(
+                reinterpret_cast<char*>(
+                    message->data), message->size);
+
+            boost::property_tree::ptree input_tree;
+
+            try
+            {
+                boost::iostreams::stream<
+                    boost::iostreams::array_source> input_stream(
+                        input.c_str(), input.size());
+
+                boost::property_tree::read_json(
+                    input_stream, input_tree);
+            }
+            catch(std::exception& error)
+            {
+                LOG_DEBUG(LOG_SERVER)
+                    << "Ignoring invalid incoming data " << error.what();
+
+                break;
+            }
+
+            const auto command = input_tree.get<std::string>(
+                "command");
+
+            if (command.find("query ") != std::string::npos)
+            {
+                const auto arguments = input_tree.get<std::string>(
+                    "arguments");
+
+                auto web_sockets =
+                    static_cast<bc::server::web_sockets*>(
+                        connection->user_data);
+
+                web_sockets->add_query_work(
+                    connection, command, arguments);
```

Comment:
Parameters are being inconsistently wrapped. Generally want to extend to 80 chars.